File: Leetcode100/leetcode100_python/BestTimetoBuyandSellStockwithCooldown.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        buy, sell, pre_buy, pre_sell = 0, 0, 0, 0
        for i in range(len(prices)):
            if i == 0:
                buy = -prices[i]
            else:
                pre_buy = buy
                buy = max(pre_buy, pre_sell - prices[i])
                pre_sell = sell
                sell = max(pre_sell, pre_buy + prices[i])
        return sell

if __name__ == "__main__":
    prices = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.maxProfit(prices))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/best_time_to_buy_and_sell_stock_with_cooldown/BestTimetoBuyandSellStockwithCooldown.java
package main.java.com.jiang.leetcode100.best_time_to_buy_and_sell_stock_with_cooldown;

import java.util.Scanner;

class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if(n == 0) return 0;

        int[] buy = new int[n];
        int[] sell = new int[n];

        for(int i = 0; i < n; i++) {
            if(i == 0) buy[i] = -prices[i];
            else {
                buy[i] = Math.max((i > 1 ? sell[i - 2] : 0) - prices[i], buy[i - 1]);
                sell[i] = Math.max(buy[i - 1] + prices[i], sell[i - 1]);
            }
        }
        return sell[n - 1];
    }
}

public class BestTimetoBuyandSellStockwithCooldown {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] prices = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) prices[i] = Integer.parseInt(nums_str[i]);

        System.out.println(new Solution().maxProfit(prices));
    }
}